Add support for multirow tooltip title

ajdani opened the pull request:
This PR adds support for text wrap of the tooltip title - currently the title + current date would just overlap the tooltip, but now the existing textWrap functionality is used for the title as well.

## Motivation and Context
Long tooltip titles + date formats break the style of the tooltip, so the solution is a text wrap for the title row as well.

## How Has This Been Tested?
- Test if the tooltipDivider changes the y-Position in case of a long title and check if the title is correct wrapped

## Screenshots (if appropriate):
Before:
![image](https://user-images.githubusercontent.com/16606530/79695416-0c75fe80-8277-11ea-9acd-1a8c77ea7f17.png)

After:
![image](https://user-images.githubusercontent.com/16606530/79695392-f0725d00-8276-11ea-8a85-675a5bdfe106.png)

## Types of changes
- [ ] Refactor (changes the way we code something without changing its functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] Latest master code has been merged into this branch
- [x] No commented out code (if required, place // TODO above with explanation)
- [ ] No linting issues
- [x] Build is successful
- [ ] Code follows the [API Guidelines](http://eventbrite.github.io/britecharts/topics-index.html#toc5__anchor)
- [ ] Updated the documentation
- [x] Added tests to cover changes
- [x] All new and existing tests passed

ajdani commented:
Oh my god, this is a flaky test - will fix it in the next days. Runs locally - seems to be dependent to the resolution of the Browser used in tests... :-(

Golodhros commented:
I see, sometimes it is useful to run 'yarn run test:travis' on dev to make sure is not something related to the environment!

Golodhros commented on src/charts/tooltip.js:
```diff
@@ -400,15 +405,68 @@ define(function(require){
+         * @return  {Number}  approximateLineNumber  Approximative number of lines needed to display the title
+         * @private
+         */
+        function getApproximateNumberOfLines() {
```
Maybe if you could make this function more generic we could add it to our text helper?

ajdani replied:
Moved it and added a test for it.

Golodhros commented on test/specs/tooltip.spec.js:
```diff
@@ -122,6 +122,57 @@ define(['jquery', 'd3', 'tooltip'], function($, d3, tooltip) {
+                        expect(parseInt(actualDividerY1Position)).toEqual(expectedDividerYPosition);
+                        expect(parseInt(actualDividerY2Position)).toEqual(expectedDividerYPosition);
+                    });
                 });
```
Thanks for the test!

ajdani commented:
Rebased and moved the function to helpers and added a test for it.

> I see, sometimes it is useful to run 'yarn run test:travis' on dev to make sure is not something related to the environment!

Thanks for the hint. Somehow this runs without errors locally. Will see if the pipeline fails here.

ajdani commented:
Ok, pipeline fails - maybe we can have a look on this together if the results of `yarn run test:travis` should be deterministic we should fix the behavior of this rather than my failing test. @Golodhros @DalerAsrorov 

I'm running `node`, `npm` and `yarn` in newer version as the CI - do you think this may cause the differences? Maybe different resolved dependencies...?

````
node --version
v12.13.0
npm --version
6.12.0
yarn --version
1.19.1
```

DalerAsrorov commented on src/charts/helpers/text.js:
```diff
@@ -150,9 +150,41 @@ define(function(require) {
+    const getApproximateNumberOfLines = function(title, fontSize, maxLength) {
+        const words = title.split(/\s+/).reverse();
+
+        var word,
```
Sorry didn't see it before, please use `let`

DalerAsrorov commented on src/charts/helpers/text.js:
```diff
@@ -150,9 +150,41 @@ define(function(require) {
+     * @private
+     */
+    const getApproximateNumberOfLines = function(title, fontSize, maxLength) {
+        const words = title.split(/\s+/).reverse();
```
is `title` always going to be defined?

ajdani replied:
Would make sense, as it's an input without a default value. Doesn't make sense to use this function without `title`.

DalerAsrorov commented on src/charts/helpers/text.js:
```diff
@@ -150,9 +150,41 @@ define(function(require) {
+            line = [],
+            approximateLineNumber = 1;
+
+        while (word = words.pop()) {
```
Why not use a simpler for loop here?
```js
for (word in words) {
// stuff
}
```

ajdani replied:
Yes, was sticking to the code-base. But changed it.

DalerAsrorov commented on src/charts/tooltip.js:
```diff
@@ -401,15 +406,45 @@ define(function(require){
+                tooltipMaxTitleLength
+            );
+
+            if(approximateNumberOfTitleLines > 1) {
```
a little bit of spacing needed here: 
```js
if (approximateNumberOfTitleLines > 1) {
...
```

DalerAsrorov commented on src/charts/tooltip.js:
```diff
@@ -401,15 +406,45 @@ define(function(require){
          * @private
          */
         function updateTitle(dataPoint) {
+            const tTitle = getTooltipTitle(dataPoint[dateLabel]);
```
What does `tTitle` represent? I think a bit more descriptive name for this would be helpful for the reader. Is it `textTitle`?

ajdani replied:
Tbh I don't know what it stands for, just sticked to the naming which was there before - so I did not introduce this variable. But it makes sense to leave it a little bit cleaner as it was before, changed it to `textTitle`, because most likely that's the meaning.

ajdani commented:
Thanks for the feedback @DalerAsrorov !

Went over it and refactored the stuff mentioned.

ajdani commented:
Ok, Travis-CI strikes back :)
Now we have the issue, that `yarn run test:travis` runs locally, but Travis-CI does not like my test. 


Maybe we can have a look on this together if the results of yarn run test:travis should be deterministic we should fix the behavior of this rather than my failing test. @Golodhros @DalerAsrorov

I'm running node, npm and yarn in newer version as the CI - do you think this may cause the differences? Maybe different resolved dependencies...?
```
node --version
v12.13.0
npm --version
6.12.0
yarn --version
1.19.1
```

Golodhros commented on src/charts/helpers/text.js:
```diff
@@ -150,9 +150,40 @@ define(function(require) {
+     * @return  {Number}  approximateLineNumber  Approximative number of lines needed to display the title
+     * @private
+     */
+    const getApproximateNumberOfLines = function(title, fontSize, maxLength) {
```
Could you set the default fontsize to 'defaultTextSize' like we do on the other helpers on this file?

Thank!

Golodhros commented on src/charts/helpers/text.js:
```diff
@@ -150,9 +150,40 @@ define(function(require) {
+        for(let word of words) {
+            line.push(word);
+
+            const textWidth = getTextWidth(line.join(' '), fontSize, 'Karla, sans-serif');
```
We probably want to add to the signature fontFace = defaultFontFace, like we do in the previous function and use the fontFace here

Golodhros commented:
Interesting, I actually get an error on your test when running the test:travis script on my machine.

This is what I have:
node 12.2.0
npm 6.9.0
yarn 1.19.2

ajdani commented:
Thanks. I will have a look on the failing test again, last time I could not reproduce it - but will check out your versions and fix it when I find the time for it!

ajdani commented:
Hi @Golodhros !

Was on this topic again, the test *does not* fail on my Windows machine, but *it fails* on my Ubuntu machine - the only difference seems to be the version of PhantomJS:
`PhantomJS 2.1.1 (Linux 0.0.0)`
vs.
`PhantomJS 2.1.1 (Windows 8.0.0)`

I would suggest to change the test to this, or even remove the check for the text of the tooltip in there: 
`it('should be displayed in two rows', () => {
                        // the space between 'Tooltip' and 'title' dissappears because of the text wrap
                        const expectedTitle = 'SuperlongandexcitingTooltiptitle-Aug05,2015';
                        const expectedDividerYPosition = 48;
                        let actualTitle, actualDividerY1Position, actualDividerY2Position;

                        tooltipChart.dateFormat(tooltipChart.axisTimeCombinations.DAY_MONTH);
                        tooltipChart.update({
                            date: '2015-08-05T07:00:00.000Z',
                            topics: []
                        }, topicColorMap, 0);

                        actualTitle = containerFixture.select('.britechart-tooltip')
                            .selectAll('.tooltip-title')
                            .text();

                        actualDividerY1Position = containerFixture.select('.britechart-tooltip')
                            .select('.tooltip-divider').attr('y1');

                        actualDividerY2Position = containerFixture.select('.britechart-tooltip')
                            .select('.tooltip-divider').attr('y2');

                        expect(actualTitle.split(" ").join("")).toEqual(expectedTitle);
                        expect(parseInt(actualDividerY1Position)).toEqual(expectedDividerYPosition);
                        expect(parseInt(actualDividerY2Position)).toEqual(expectedDividerYPosition);
                    });
`

Golodhros commented:
Hi @ajdani, thanks for following up!

Go for it! You can do whatever you feel necessary to make that test pass.

For the record, we are probably going to get rid of PhantomJS, as it doesn't work great with Github actions. Not sure when, but soon!

Thanks for your contribution!

Golodhros commented:
@all-contributors please add @ajdani for bug, code, maintenance

Golodhros commented:
Merged as https://github.com/britecharts/britecharts/pull/849
